internal/vault_reader.go

```diff
@@ -0,0 +1,183 @@
+	}
+
+	if secret.Data == nil {
+		return nil, errors.New("returned Vault secret has a nil Data map")
+	}
+
+	if len(secret.Data) == 0 {
+		return nil, errors.New("returned Vault secret has an empty Data map")
+	}
+
+	if secret.Data["data"] == nil {
+		return nil, fmt.Errorf("returned Vault secret has a Data map that contains no value for key 'data'. Available keys are: %v", availableKeys(secret.Data))
+	}
+
+	data, ok := secret.Data["data"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("data type assertion failed for Vault secret of type: %T and value %#v read from path %s", secret.Data["data"], secret.Data["data"], path)
+	}
+
+	username, err := getValue("username", data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get username from Vault secret: %w", err)
+	}
+
+	password, err := getValue("password", data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get password from Vault secret: %w", err)
+	}
+
+	return ClusterCredentials{username: username, password: password}, nil
+}
+
+func getValue(key string, data map[string]interface{}) (string, error) {
+	result, ok := data[key].(string)
+	if !ok {
+		return "", fmt.Errorf("expected %s to be a string but is a %T", key, data[key])
+	}
+
+	return result, nil
+}
+
+func availableKeys(m map[string]interface{}) []string {
+	result := make([]string, len(m))
+	i := 0
+	for k := range m {
+		result[i] = k
+		i++
+	}
+	return result
+}
+
+func InitializeSecretStoreClient(vaultSpec *rabbitmqv1beta1.VaultSpec) (SecretStoreClient, error) {
+	// GCH TODO return to this...
+	// role := vaultSpec.Role
+	// if role == "" {
+	// 	return nil, errors.New("no role value set in Vault secret backend")
+	// }
+	role := "messaging-topology-operator"
+
+	// For now, the VAULT_ADDR environment variable will be the address that your pod uses to communicate with Vault.
+	config := vault.DefaultConfig() // modify for more granular configuration
+
+	vaultClient, err := vault.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize Vault client: %w", err)
+	}
+
+	var annotations map[string]string = vaultSpec.Annotations
+	if annotations["vault.hashicorp.com/namespace"] != "" {
+		vaultClient.SetNamespace(annotations["vault.hashicorp.com/namespace"])
+	}
+
+	jwt, err := ServiceAccountTokenReader()
+	if err != nil {
+		return nil, fmt.Errorf("unable to read file containing service account token: %w", err)
+	}
+
+	loginAuthPath := defaultAuthPath
+	if annotations["vault.hashicorp.com/auth-path"] != "" {
+		loginAuthPath = annotations["vault.hashicorp.com/auth-path"]
+	}
+
+	vaultToken, err := VaultClientTokenReader(vaultClient, string(jwt), role, loginAuthPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read Vault client token: %w", err)
+	}
+
+	// use the Vault token for making all future calls to Vault
+	vaultClient.SetToken(vaultToken)
+
+	return VaultClient{Reader: &VaultSecretReader{client: vaultClient}}, nil
```

Comment:
Have you considered caching the VaultClient or the `vaultToken`? I would imagine authentication is going to be more expensive if we compared it looking up a secret's path. I guess we could cache the vaultToken or the actual initialized VaultClient and reuse it over and over again until the token expires. 

I am concerned with the messaging-topology-operator overloading Vault if it has to reconcile way too many resources. 

This does not need to be done now. But it is worth considering for future releases.

Reply:
I would second what @MarcialRosales wrote: since applications will typically create hundreds to thousands of queues and bindings it is well worth caching the vault session

Reply:
As a strawman, I've made a change to `vault_reader.go` so that after the initial login to Vault the returned token is periodically renewed in a background thread. The maxiumum TTL of the returned token is also respected so that when reached, a new Vault login is triggered to retrieve a fresh token for subsequent renewals. Would you be happy with this kind of approach?

Reply:
@georgeharley Sorry I did not reply directly to your question. Yes, I am happy with your approach and in fact it adheres to Vault Go Client library recommendation. I am happy to leave the TTL up to what the server says rather providing one.